File: cpp/chapter2/fence_repair2.cpp
/* Fence Repair
 * 順位キュー
 */

#include <iostream>
#include <queue>

using namespace std;

typedef long long ll;

int main(){
  int N;
  cin >> N;
  int L[N];
  for(int i = 0; i < N; i++){
    cin >> L[i];
  }

  ll ans = 0;

  // 順位キューを用意（小さい数から出てくるようにする
  priority_queue<int, vector<int>, greater<int> > que;
  for(int i = 0; i < N; i++){
    que.push(L[i]);
  }

  // 枝が１本になるまで適用
  while(que.size() > 1){
    // 1番短い板、次に短い板を取り出す
    int l1, l2;
    l1 = que.top();
    que.pop();
    l2 = que.top();
    que.pop();

    // それらを結合
    ans += l1 + l2;
    que.push(l1 + l2);
  }

  cout << ans << endl;

  return 0;
}

File: python/chapter2/fence_repair2.py
"""
大きな板からN個の板を切り出す。
それらの長さはL_1, L_2, ..., L_Nであり、元の板はそれらの板の長さの合計となる。
板を切る時はその長さ分コストがかかる。
全ての板を切り出すのに、最小のコストはいくつか

3
8 5 8
=>34
"""
import queue

n = int(input())
L = list(map(int, input().split()))

def solve_que():
    """
    順位キューを使って解く。

    Returns
    ----------
    ans : int
        最小のコスト
    """
    ans = 0
    que = queue.PriorityQueue()
    for l in L:
        que.put(l)

    while que.qsize() > 1:
        l1 = que.get()
        l2 = que.get()
        ans += l1 + l2
        que.put(l1+l2)

    return ans

ans = solve_que()
print(ans)
